### PR TITLE
Fixed #27998 -- Made ManyToManyField changes logged in admin's object history.

### PR DESCRIPTION
**Pull request by ljsjl**
Added call to form.has_changed() in ModelAdmin._changeform_view() before
self.save_related() is called. This ensures querysets in form.initial
are evaluated before m2m field changes are committed.
ModelAdmin.changelist_view() already checks form.has_changed() first so
no change required.

### CODE REVIEW

File: tests/admin_views/tests.py [hunk 1]
```diff
@@ -6181,3 +6181,36 @@ def test_implicitly_generated_pk(self):
...
+                'date_joined_0': str(user.date_joined.date()),
+                'date_joined_1': str(user.date_joined.time().strftime('%X')),
+                'initial-date_joined_0': str(user.date_joined.date()),
+                'initial-date_joined_1': str(user.date_joined.time().strftime('%X')),
+                '_save': '1'
+                }
+        self.client.post(reverse('admin:auth_user_change', args=(user.pk,)), post_data)
+        user_ctype = ContentType.objects.get_for_model(User)
+        log = LogEntry.objects.filter(content_type=user_ctype, object_id=user.pk).last()
+        self.assertEqual(log.get_change_message(), 'Changed groups.')
```
Comment (charettes) [inline tests/admin_views/tests.py]: Thank you for your contribution! Adding a test to make sure no message is logged if nothing changed in an inline would also be required.

**Comment by ljsjl** [comment]
Added the test, though I wasn't totally sure what situation it is meant to  test so evidently I didn't quite understand what you were after.

File: tests/admin_views/models.py [hunk 2]
```diff
@@ -948,3 +948,13 @@ def __str__(self):
...
 class RelatedWithUUIDPKModel(models.Model):
     parent = models.ForeignKey(ParentWithUUIDPK, on_delete=models.SET_NULL, null=True, blank=True)
+
+
+# Models for #27998
+class Fooble(models.Model):
+    name = models.CharField(max_length=20)
+
+
+class Warble(models.Model):
```
Comment (timgraham) [inline tests/admin_views/models.py]: Could you please reuse an existing model with a `ManyToManyField`?

File: tests/admin_views/tests.py [hunk 3]
```diff
@@ -6181,3 +6182,46 @@ def test_implicitly_generated_pk(self):
         post_data = {'name': '2'}
         response = self.client.post(reverse('admin:admin_views_implicitlygeneratedpk_change', args=(1,)), post_data)
         self.assertEqual(response.status_code, 302)
+
+
+@override_settings(ROOT_URLCONF='admin_views.urls')
+class AddM2MFieldsToObjChangeHistoryTest(TestCase):
```
Comment (timgraham) [inline tests/admin_views/tests.py]: See if you can find an existing test case to use. There should be existing tests for log entries that this test could be located with.

Comment (ljsjl) [inline tests/admin_views/tests.py]: Aren't many tests for log entries in admin_views module, but  test_change_view in admin_views/tests.py looks the best place to add this test.

 I could move it to the admin_utils module which has all the logentry tests, though it's not really a logentry test more a statement order check in view processing so doesn't feel quite right to me but I'll defer. I'd need to create a couple of models as there are none with m2m fields in that module but it's a small module, so maybe better than cluttering up admin_views?

Comment (timgraham) [inline tests/admin_views/tests.py]: `admin_views` is the place for the test since this is testing a change in the admin views. Putting it near `test_change_view` sounds reasonable.

File: tests/admin_views/tests.py (hunk 1 above)
Comment (timgraham) [inline tests/admin_views/tests.py]: @charettes, could you explain your reasoning here? This scenario is already working without the patch -- perhaps the additional test coverage can be added separately to make this patch simpler. The only assertion that fails currently is `self.assertEqual(log.get_change_message(), 'Changed has_fooled_today.')`. I'm not sure if "inline" in your comment might be a typo for `ManyToManyField`?

Comment (charettes) [inline tests/admin_views/tests.py]: @timgraham It might be more appropriate in another commit then. I believe I wanted to make sure nothing was logged if a m2m backed inline was submitted without changes.

**Comment by wonderbeyond** [comment]
Same issue in 1.11.x.

**Comment by timgraham** [comment]
It doesn't qualify for a backport to 1.11 -- see our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions).

**Comment by wonderbeyond** [comment]
So normal bug fixes only stay in master for next feature relaese branch?

**Comment by timgraham** [comment]
Correct. However, as I subsequently commented in the ticket for this issue, I noticed that this is a regression in Django 1.10 so it's okay to backport a fix. See https://github.com/django/django/pull/8993.
